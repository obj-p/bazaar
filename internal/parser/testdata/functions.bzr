package functions

data FunctionFields {
        empty func()
        add func(int, int) -> int
        trailingComma func(int,)
        optional func()?
        nestedOptional (func() -> string?)?
        extraNested (((func()?)))
}

func HigherOrderFunction() -> func() -> func()
