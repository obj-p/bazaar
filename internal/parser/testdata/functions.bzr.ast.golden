&parser.Bazaar{
  Package: &parser.PackageDecl{
    Domain: []string{
      "functions",
    },
  },
  Decls: []*parser.TopLevelDecl{
    {
      Data: &parser.DataDecl{
        Name: "FunctionFields",
        Members: []*parser.MemberDecl{
          {
            Field: &parser.FieldDecl{
              Name: "empty",
              Type: &parser.TypeDecl{
                Function: &parser.FunctionTypeDecl{
                },
              },
            },
          },
          {
            Field: &parser.FieldDecl{
              Name: "add",
              Type: &parser.TypeDecl{
                Function: &parser.FunctionTypeDecl{
                  Parameters: []*parser.TypeDecl{
                    {
                      Value: &"int",
                    },
                    {
                      Value: &"int",
                    },
                  },
                  Return: &parser.TypeDecl{
                    Value: &"int",
                  },
                },
              },
            },
          },
          {
            Field: &parser.FieldDecl{
              Name: "trailingComma",
              Type: &parser.TypeDecl{
                Function: &parser.FunctionTypeDecl{
                  Parameters: []*parser.TypeDecl{
                    {
                      Value: &"int",
                    },
                  },
                },
              },
            },
          },
          {
            Field: &parser.FieldDecl{
              Name: "optional",
              Type: &parser.TypeDecl{
                Function: &parser.FunctionTypeDecl{
                },
                Optional: true,
              },
            },
          },
          {
            Field: &parser.FieldDecl{
              Name: "nestedOptional",
              Type: &parser.TypeDecl{
                Nested: &parser.TypeDecl{
                  Function: &parser.FunctionTypeDecl{
                    Return: &parser.TypeDecl{
                      Value: &"string",
                      Optional: true,
                    },
                  },
                },
                Optional: true,
              },
            },
          },
          {
            Field: &parser.FieldDecl{
              Name: "extraNested",
              Type: &parser.TypeDecl{
                Nested: &parser.TypeDecl{
                  Nested: &parser.TypeDecl{
                    Function: &parser.FunctionTypeDecl{
                    },
                    Optional: true,
                  },
                },
              },
            },
          },
        },
      },
    },
    {
      Function: &parser.FunctionDecl{
        Name: "HigherOrderFunction",
        Return: &parser.TypeDecl{
          Function: &parser.FunctionTypeDecl{
            Return: &parser.TypeDecl{
              Function: &parser.FunctionTypeDecl{
              },
            },
          },
        },
      },
    },
  },
}
